client/with_dry_runs.go

```diff
@@ -21,133 +21,117 @@ import (
 	"github.com/mysteriumnetwork/payments/bindings"
 	"github.com/pkg/errors"
 )
 
+type estimatable interface {
+	GetGasLimit() uint64
+	toEstimator(ethClientGetter) (*bindings.ContractEstimator, error)
+	toEstimateOps() *bindings.EstimateOpts
+}
+
+// ErrorTransactionReverted when contract execution is interrupted with an error./
+type ErrorTransactionReverted struct {
+	Err    rpc.Error
+	Reason string
+}
+
+func (e ErrorTransactionReverted) Error() string {
+	return e.Err.Error()
+}
+
 // WithDryRuns forces a dry run before running a write transaction on blockchain.
 // Ethereum client will perform a dry run on a transaction with no gas limit set.
 // This component will perform a dry run if and only if the gas limit is set to a non zero value.
 // In this way, the dry run is always performed before sending the transaction to the network.
 // For convenience, this component proxies read only calls to the underlying blockchain.
 type WithDryRuns struct {
-	bc blockchain
+	bc        blockchain
+	ethClient ethClientGetter
 }
 
 // NewWithDryRuns creates a new instance of client with dry runs.
-func NewWithDryRuns(bc blockchain) *WithDryRuns {
+func NewWithDryRuns(bc blockchain, ethClient ethClientGetter) *WithDryRuns {
 	return &WithDryRuns{
-		bc: bc,
+		bc:        bc,
+		ethClient: ethClient,
 	}
 }
 
-type gasLimitProvider interface {
-	GetGasLimit() uint64
-}
-
-// The params for the dry run should be in the strict order as required by the smart contract.
-func (cwdr *WithDryRuns) dryRun(glp gasLimitProvider, binding string, from, to common.Address, method string, params ...interface{}) (uint64, error) {
+// Estimate estimates the (paid) contract gas price.
+func (cwdr *WithDryRuns) Estimate(req estimatable) (uint64, error) {
 	// If the gas limit is set to 0, ethereum client will do the estimation for us.
 	// We only force the estimation if the gas limit is set to a non zero value.
-	if glp.GetGasLimit() == 0 {
+	if req.GetGasLimit() == 0 {
 		return 0, nil
 	}
-	parsed, err := abi.JSON(strings.NewReader(binding))
-	if err != nil {
-		return 0, errors.Wrap(err, "could not deserialize ABI")
-	}
-	input, err := parsed.Pack(method, params...)
+
+	estimator, err := req.toEstimator(cwdr.ethClient)
 	if err != nil {
-		return 0, errors.Wrap(err, "could not pack input")
+		return 0, err
 	}
-	msg := ethereum.CallMsg{From: from, To: &to, Data: input}
-	gas, err := cwdr.bc.EstimateGas(msg)
+
+	gas, err := estimator.Estimate(req.toEstimateOps())
 	return gas, errors.Wrap(err, "could not estimate gas")
 }
 
+// DryRun simulates the (paid) contract method with params as input values.
+func (cwdr *WithDryRuns) DryRun(req estimatable) error {
+	_, err := cwdr.Estimate(req)
+	if err == nil {
+		return nil
+	}
+
+	// Extract error thrown in contract
+	if rpcCauseErr, hasCause := errors.Cause(err).(rpc.Error); hasCause {
+		if rpcCauseErr.ErrorCode() == -32000 && strings.Contains(rpcCauseErr.Error(), "VM Exception while processing transaction: revert") {
```

Comment:
Im gonna complain without a solution here but could we describe this better? Move it to another func? Or somehow change it so it reads better?

Reply:
Validate() is already special function for that